prow/pod-utils/gcs/upload.go

```diff
@@ -266,28 +230,43 @@ type dataWriter interface {
 	io.WriteCloser
 	fullUploadPath() string
 	ApplyWriterOptions(opts pkgio.WriterOptions)
-	compressData() bool
 }
 
 type openerObjectWriter struct {
 	pkgio.Opener
-	Context  context.Context
-	Bucket   string
-	Dest     string
-	compress bool
-	opts     []pkgio.WriterOptions
-	writer   pkgio.Writer
-	closers  []pkgio.Closer
+	Context          context.Context
+	Bucket           string
+	Dest             string
+	compressFileType bool
+	opts             []pkgio.WriterOptions
+	writer           pkgio.Writer
+	closers          []pkgio.Closer
 }
 
 func (w *openerObjectWriter) Write(p []byte) (n int, err error) {
 	if w.writer == nil {
+		largerThanOneKB := len(p) > 1024
+		alreadyCompressed := http.DetectContentType(p) == "application/x-gzip"
```

Comment:
nit: this reads 512 bytes, when multiplies by 1k files you start to see an impact. I appreciate the named booleans but in-lining this into the conditional would defer the write until it was actually necessary.